### PR TITLE
doc: add CSP examples

### PR DESCRIPTION
**Pull request by zeke**
This PR adds some examples of how to configure a Content Security Policy, using either an HTTP header or a `<meta>` tag.

🍐 @liliakai

### CODE REVIEW

**Comment by zeke** [comment]
Would love to get input on a sensible default policy. `default-src 'self'` is probably too restrictive for most apps. But maybe the default example should be pretty strict?

@sindresorhus does your example in #12909 represent what you think are good defaults?

File: docs/tutorial/security.md [hunk 1]
```diff
@@ -357,6 +352,32 @@ Content-Security-Policy: '*'
...
+[`webRequest.onHeadersReceived`](../api/web-request.md#webrequestonheadersreceivedfilter-listener)
+handler:
+
+```javascript
+const {session} = require('electron')
+
+session.defaultSession.webRequest.onHeadersReceived((details, callback) => {
+  callback({responseHeaders: `default-src 'self'`})
+})
+```
```
Comment (sindresorhus) [inline docs/tutorial/security.md]: I think you should point out that the best would be to start with `default-src 'none';` and then allow what's needed. It's always better to include than exclude. Most will probably then end up with:

```
default-src 'none';
script-src 'self';
img-src 'self' data:;
style-src 'self';
font-src 'self';
```

Comment (felixrieseberg) [reply docs/tutorial/security.md]: That's a solid point, I agree with that.

Comment (zeke) [reply docs/tutorial/security.md]: Updated to `default-src 'none'`

**Comment by sindresorhus** [comment]
I think you should also include these by in the example as they're not covered by `default-src 'none';`:

```
base-uri 'none';
form-action 'none';
frame-ancestors 'none';
```

I would also link to https://developers.google.com/web/fundamentals/security/csp/ which is another good resource on this topic.

**Comment by furrykef** [comment]
I don't find the current examples sufficient. For example, I have no idea where the "session.defaultSession.webRequest.onHeadersReceived" magic incantation is supposed to go. The "Writing Your First Electron App" page also does not even mention this issue despite the fact such an app will display a warning about CSP. Finally, since I haven't used CSP in any form at all before, let alone with Electron, it'd be nice to have a guide for what policies to set.

File: docs/tutorial/security.md [hunk 2]
```diff
@@ -357,6 +352,32 @@ Content-Security-Policy: '*'
...
+Electron respects the [`Content-Security-Policy` HTTP header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy)
+which can be set using Electron's 
+[`webRequest.onHeadersReceived`](../api/web-request.md#webrequestonheadersreceivedfilter-listener)
+handler:
+
+```javascript
+const {session} = require('electron')
+
+session.defaultSession.webRequest.onHeadersReceived((details, callback) => {
+  callback({responseHeaders: `default-src 'none'`})
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply docs/tutorial/security.md]: // @zeke @felixrieseberg @MarshallOfSound :arrow_up:

[the remark this answers is not in the document]
Comment (bcomnes) [reply docs/tutorial/security.md]: 👍

**Comment by justinmeiners** [comment]
Is `onHeadersReceived` called by `file://` protocol URLs? If not, why would that be the recommended way of setting up CSP?

**Comment by Slapbox** [comment]
Can we possibly improve this example? I and others have spent a lot of time trying to figure out how to get this working. None of the answers to this question on StackExchange seem to work for anyone but the submitter.

https://stackoverflow.com/questions/51969512/define-csp-http-header-in-electron-app

**Comment by smariel** [comment]
> Is `onHeadersReceived` called by `file://` protocol URLs? If not, why would that be the recommended way of setting up CSP?

As @justinmeiners said, it seems that `onHeadersReceived` is not called by `file://` protocol, unlike `onResponseStarted`. But `onResponseStarted` do not provides a callback object, so it is not possible to add CSP.
   
    app.on('ready', () => {
      session.defaultSession.webRequest.onHeadersReceived((details,callback) => {
        console.log(`onHeadersReceived:\n${details.url}`);
      });
      session.defaultSession.webRequest.onResponseStarted((details) => {
        console.log(`onResponseStarted:\n${details.url}`);
      });
      ...
    });

**Comment by busticated** [comment]
adding my voice to those asking for clearer docs and examples. the only thing i've been able to get working is this 👉 https://stackoverflow.com/a/52243138/579167

more problematically, while this approach appears to work as expected when i'm running in _development_ (where my `viewURL` is an `http` url), when i run my packaged app (where my `viewURL` uses the `file://` protocol), previously blocked content is rendered and visible (i'm setting `img-src 'none'` for testing so it's _really_ easy to tell)

i've tried:
* the [example in the docs](https://electronjs.org/docs/tutorial/security#csp-http-header)
* setting `extraHeaders ` in my `mainWindow.loadURL()` call ([docs](https://electronjs.org/docs/api/browser-window#winloadurlurl-options))
* setting it via the `<meta http-equiv="Content-Security-Policy">` tag

of those the `meta` tag _actually_ works but doesn't support the `frame-ancestors` directive (and has a few other drawbacks unique to my use-case).

any help here would be very much appreciated 🙏 

cc @zeke @felixrieseberg

**Comment by MarshallOfSound** [comment]
@busticated The `file://` protocol does not support or recognize HTTP headers (because it is not an HTTP protocol).  If you want to set a CSP on a `file://` protocol file you need to use the `meta` tag strategy for setting a CSP.

**Comment by smariel** [comment]
@MarshallOfSound thanks for this answer. So there is no way to set CSP in the main process if renderers are local files packaged in the application.

Do you know why `details.headers` and `details.responseHeaders` exist eaven with file:// protocol, when using `onResponseStarted()` ?

**Comment by justinmeiners** [comment]
@MarshallOfSound Thank you for this explanation. Since most Electron apps use local resources (file: protocol) shouldn't meta tag be the recommended way of doing CSP? 

At the very least can we update the documentation to explain when to use "meta" tag instead of discouraging it?

**Comment by busticated** [comment]
thanks @MarshallOfSound 👍

it seems like that pattern - using `http:` for "development", `file:` for the package app - which i've seen fairly often out in the wild is likely to cause all sorts of weird edges cases. for example 👉 https://github.com/whatwg/html/issues/3099

i realize this is somewhat off-topic but what's the guidance there? do most professional (🙄) apps use `file://` for everything (or vice-versa)? (looks like atom [uses `file://`](https://github.com/atom/atom/blob/master/src/main-process/atom-window.js#L124))

at the least, probably worth flagging in the docs along with calling out that the CSP headers should _only_ be set for `Content-Type: text/html` (as well as `application/xhtml+xml` and `application/xml` i suppose) responses.

**Pull request merged**